sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReader.java

```diff
@@ -0,0 +1,148 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.cloud.bigquery.v3.ParallelRead.ReadLocation;
+import com.google.cloud.bigquery.v3.ParallelRead.ReadOptions;
+import com.google.cloud.bigquery.v3.ParallelRead.ReadRowsRequest;
+import com.google.cloud.bigquery.v3.ParallelRead.ReadRowsResponse;
+import com.google.cloud.bigquery.v3.ParallelRead.Session;
+import com.google.cloud.bigquery.v3.RowOuterClass.Row;
+import java.io.IOException;
+import java.util.Iterator;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Iterates over all rows in a table.
+ */
+class BigQueryParallelReader<T> extends BoundedSource.BoundedReader<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReader.class);
+
+  private Session session;
+  private final BigQueryServices client;
+  private Iterator<ReadRowsResponse> responseIterator;
+  private Iterator<Row> rowIterator;
+  private final ReadRowsRequest request;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private BoundedSource<T> source;
+  private BigQueryOptions options;
+  private Row currentRow;
+
+  private BigQueryParallelReader(
+      Session session,
+      ReadLocation readLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    this.session = checkNotNull(session, "session");
+    this.client = checkNotNull(client, "client");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.source = checkNotNull(source, "source");
+    this.options = options;
+
+    ReadRowsRequest.Builder builder = ReadRowsRequest.newBuilder().setReadLocation(readLocation);
+    if (rowBatchSize != null) {
+      builder.setOptions(ReadOptions.newBuilder().setMaxRows(rowBatchSize).build());
+    }
+
+    this.request = builder.build();
+  }
+
+  public static <T> BigQueryParallelReader<T> create(
+      Session session,
+      ReadLocation initialReadLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    return new BigQueryParallelReader<>(
+        session,
+        initialReadLocation,
+        rowBatchSize,
+        parseFn,
+        source,
+        client,
+        options);
+  }
+  /**
+   * Empty operation, the table is already open for read.
+   * @throws IOException on failure
+   */
+  @Override
+  public boolean start() throws IOException {
+    responseIterator = client.getTableReadService(options).readRows(request);
```

Comment:
Is it possible to just say `return advance();` here, or refactor the code in a form where that would be the case?

Reply:
Done.